Understanding rails forms

expjazz opened the pull request:
- In the code, there are three different ways to work with forms on Ruby On Rails.
- One using the old HTML way
- Another with the form_tag
- The last with form_for
- We have a new, create, edit, and update actions as required.
- We show messages to the user with the content of the errors if the process of creating or editing a user did not work.
- We made a list on the new_page to show all the users that were created and to give a path to edit the profiles.
- We did not style the page because this assignment is only for the purpose of understanding forms.

![image](https://user-images.githubusercontent.com/61479648/82447565-d8147e80-9a7e-11ea-9dad-c2f3aef27dcf.png)

frankopkusianwar committed:
create html form

frankopkusianwar committed:
add for_tag form

frankopkusianwar committed:
add form_for form

expjazz committed:
Added messages. Added edit form and edit view

expjazz committed:
Added readme

expjazz committed:
fixed linter errors

expjazz committed:
linter

expjazz committed:
linter errors

expjazz committed:
linter errors

kobiyoyo commented on config/routes.rb:
```diff
@@ -1,3 +1,5 @@
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  root 'users#new'
+
+  resources :users
```
let's specify only routes for the actions we used in `User` controller.

kobiyoyo commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,42 @@
+
+  private
+
+  def create_params
```
this should be named `user_params`

kobiyoyo commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,42 @@
+class UsersController < ApplicationController
+  def create
+    @user = User.new(create_params)
+    words = ''
+    # @user = User.new(username: params[:username], email: params[:email], password: params[:password])
+    if @user.save
+      flash[:success] = 'Your profile was created.'
+    else
+      @user.errors.full_messages.each do |msg|
+        words += "#{msg} "
+      end
+      flash[:error] = words
+    end
+    redirect_to new_user_path
+  end
```
if unsuccessful  `render :new`

kobiyoyo commented on app/views/users/edit.html.erb:
```diff
@@ -0,0 +1,32 @@
+<h2> Edit your profile </h2>
+
+<%# <form action="/users" method="post" accept-charset="UTF-8"> %>
```
lets use `<%# <form action= users_path method="post" accept-charset="UTF-8"> %>`

kobiyoyo commented on app/views/users/edit.html.erb:
```diff
@@ -0,0 +1,32 @@
+  <input type="submit" value="submit">
+</form> %>
+
+<%# <%= form_tag('/users', method: "post") do %> <%#%%>
```
lets use `<%# <%= form_tag(users_path, method: "post") do %> <%#%%>`

kobiyoyo commented on app/views/users/edit.html.erb:
```diff
@@ -0,0 +1,32 @@
+  <%= f.label :email %>
+  <%= f.email_field :email %>
+  <%= f.label :password %>
+  <%= f.password_field :email %>
```
this should be ` <%= f.password_field  :password %>`

frankopkusianwar replied:
## We worked on all the changes below

- we replaced the redirect_to with render :new after failing to create a user
- we replaced the action in the forms with user_path instead of "/users"
- we replace create_params with user_params
- we added the only to the routes resources for specific actions in the controller
- we replaced the email on password field with password
- we also added validations for password

kobiyoyo reviewed (changes requested):
## Hi Team      ![](https://user-images.githubusercontent.com/38801423/62908122-703d4480-bd44-11e9-9115-42025ab7f341.png) 
#### Great Job so far, but there are a few more things required. 
### General
- [ ] Lets add validation for `User` password.
- [ ] check comments below

### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

frankopkusianwar committed:
fix pull request issues

Forison reviewed (changes requested):
## Status:: Require changes ❌ 
# 👋 team,
you have done a good job so far, however, some of the issues raised in the previous review still persist.
* [ ] ```resources :users, only: [:new, :create, :edit, :update]``` in config/routes.rb ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_route1.png) [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_route2.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/forms))
- Currently ```post 'users/new', to: 'users#create'``` is not necessary as it duplicates what the above does.
* [ ] Let us remove commented code from the users_controller in lines 5 and 20
* [ ] Let us remove commented codes from the view/users/new file.
* [ ] Let us remove commented codes from the view/users/edit file.
* [ ] Show profile link return an error since it was not implemented. Make sure all the links are working else don't avail them.
> Having clean code is a best practice as a developer🌝.
## Kindly make these changes and submit the project for another review.
#### Happy coding💻
Addo

expjazz committed:
change as requested by the TSE

expjazz commented:
### We made the requested changes listed above:
- We removed that unnecessary route;
- We removed the comments
- We removed the link to the users page, since having an user page its not a requirement.
### However, in the Odin assignment says that we should leave the other type of forms in the code as comments. Since the last TSE @Forison asked us to remove, we removed them.

mcrd25 reviewed (approved):
# Changes approved! 🎊 
Hi Team! 👋 

Yes, you are right! You were required to leave to commented out forms!

Sorry for this misunderstanding! 😅 
Everything looks good to me!

You may not merge and move on!

Good luck with the ⏭️ project! 😄 

***If you have any questions, now or in the future, do not hesitate to contact me on social media, email or on slack and I will be glad to chat with you.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)
